beluga_amcl/src/ndt_amcl_node_3d.cpp

```diff
@@ -0,0 +1,503 @@
+    return;
+  }
+
+  // Transforms are always published to keep them current.
+  if (enable_tf_broadcast_ && get_parameter("tf_broadcast").as_bool()) {
+    if (last_known_odom_transform_in_map_.has_value()) {
+      auto message = geometry_msgs::msg::TransformStamped{};
+      // Sending a transform that is valid into the future so that odom can be used.
+      const auto expiration_stamp = tf2_ros::fromMsg(laser_scan->header.stamp) +
+                                    tf2::durationFromSec(get_parameter("transform_tolerance").as_double());
+      message.header.stamp = tf2_ros::toMsg(expiration_stamp);
+      message.header.frame_id = get_parameter("global_frame_id").as_string();
+      message.child_frame_id = get_parameter("odom_frame_id").as_string();
+      message.transform = tf2::toMsg(last_known_odom_transform_in_map_.value());
+      tf_broadcaster_->sendTransform(message);
+    }
+  }
+
+  // New pose messages are only published on updates to the filter.
+  if (new_estimate.has_value()) {
+    // ROS covariances use extrinsic RPY parametrization, we need to convert to it from our tangent space
+    // representation. We use an unscented transform to apply such transform as it's non linear.
+    // See https://www.ros.org/reps/rep-0103.html#covariance-representation .
+    static const auto kTangentialSpaceToXyzRpy =
+        [](const Eigen::Vector<double, 6>& tangent) -> Eigen::Vector<double, 6> {
+      const auto se3 = Sophus::SE3d::exp(tangent);
+      // Eigen's eulerAngles uses intrinsic rotations, but XYZ extrinsic rotation is the same as ZYX intrinsic rotation.
+      // See https://pages.github.berkeley.edu/EECS-106/fa21-site/assets/discussions/D1_Rotations_soln.pdf
+      // This gives (extrinsic) yaw, pitch, roll in that order.
+      const Eigen::Vector3d euler_angles = se3.so3().matrix().eulerAngles(2, 1, 0);
+      Eigen::Vector<double, 6> ret;
+      ret.head<3>() = se3.translation();
+      ret[3] = euler_angles.z();
+      ret[4] = euler_angles.y();
+      ret[5] = euler_angles.x();
+      return ret;
+    };
+    auto message = geometry_msgs::msg::PoseWithCovarianceStamped{};
+    message.header.stamp = laser_scan->header.stamp;
+    message.header.frame_id = get_parameter("global_frame_id").as_string();
+    auto [base_pose_in_map, base_pose_covariance] = new_estimate.value();
+    tf2::toMsg(base_pose_in_map, message.pose.pose);
+
+    static constexpr double kMinimumVariance =
+        1e-6;  // Make sure we covariance is not singular so that UT doesn't fail.
+    for (auto index = Eigen::Index{0}; index <= base_pose_covariance.cols(); ++index) {
+      base_pose_covariance.coeffRef(index, index) = std::max(base_pose_covariance(index, index), kMinimumVariance);
+    }
+    const auto& [base_pose_in_map_xyz_rpy, base_pose_covariance_xyz_rpy] =
+        beluga::unscented_transform(base_pose_in_map.log(), base_pose_covariance, kTangentialSpaceToXyzRpy);
```

Comment:
@serraramiro1 I haven't reviewed the math yet, but this looks like something that should live in `beluga_ros`.